src/dataframe/dataframe.jl

```diff
@@ -1212,34 +1248,61 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
 
 Base.convert(::Type{DataFrame}, d::AbstractDict) = DataFrame(d, copycols=false)
 
-function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; columns::Symbol=:equal)
-    if !(columns in (:equal, :intersect))
-        throw(ArgumentError("`columns` keyword argument must be `:equal` or `:intersect`"))
+function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; cols::Symbol=:setequal,
+                    columns::Union{Nothing,Symbol}=nothing)
+    if columns !== nothing
+        cols = columns
+        Base.depwarn("`columns` keyword argument is deprecated. Use `cols` instead. ", :push!)
+    end
+    possible_cols = (:orderequal, :setequal, :intersect, :subset)
+    if !(cols in possible_cols)
+        throw(ArgumentError("`cols` keyword argument must be any of :" * join(possible_cols, ", :")))
     end
     nrows, ncols = size(df)
+    targetrows = nrows + 1
     if ncols == 0 && row isa NamedTuple
         for (n, v) in pairs(row)
             setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
         end
         return df
     end
-    # Only check for equal lengths, as an error will be thrown below if some names don't match
-    if columns === :equal && length(row) != ncols
-        # TODO: add tests for this case after the deprecation period
-        Base.depwarn("In the future push! will require that `row` has the same number" *
-                      " of elements as is the number of columns in `df`. " *
-                      "Use `columns=:intersect` to disable this check.", :push!)
+    if cols == :orderequal
+        if row isa NamedTuple
+            if length(row) != ncol(df) || any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
+                throw(ArgumentError("when `cols=:orderequal` all data frames must have " *
+                                    "the same column names and in the same order"))
+            end
+        end
+    end
+    # for AbstractDict :orderequal is allowed ant treated as :setequal
+    if cols == :setequal || (cols == :orderequal && row isa AbstractDict) || cols === :equal
```

Comment:
This will preclude dicts that have an order (like OrderedDict and SortedDict). There also is no easy opt-in/opt-out. I think it is a fair decision to allow failing on unordered collections if `cols=:orderequal` is supplied. I'd not default to ignoring order if it is explicitely specified (remember that `:setequal` is the default. Thus, `:orderequal` will only appear if explicitely required)

Reply:
So do you want `:orderequal` to test for equality of what is returned by `keys`? Then for `Dict` it will be kind of non-deterministic if it passes or fails for the user. We might allow this - in such a case I would add it to the docstring.

Reply:
That was what I would suggest as long as there is no function like `hasorder` which can tell us whether the structure supports a deterministic order. My reason for that is, that `:orderequal` will only be used when specified by the user. Thus, in my opinion we should use the provided order no matter whether that will work or not. Do what the user specifies as long as there are cases in which it is the expected outcome. If we knew that all `T<:AbstractDict` would be unordered, then I'd maybe consider doing that special casing. But since we don't know that, just keep doing what is told us to do. And that would mean to check the order.

Reply:
Maybe we can accept `AbstractDict` if `keys` match in the right order, but always throw an error for `Dict`? That will catch most cases and will be better than a non-deterministic error. (I agree that ideally there would be an `hasorder` trait.)

BTW, on the line above, "ant" should be "and".

Reply:
I agree this is a good trade-off. Changed

Reply:
I guess specifying that calling `:orderequal` on unordered collections (aka where pairs is non-deterministic or the order is undefined) will result in undefined behaviour. An error counts to undefined behaviour in my opinion so its just a note about documentation.
Btw would we want keys or pairs? I guess keys is already the right way because it sort of separates column specification from actual merging(?)

Reply:
I guess specifying that calling `:orderequal` on unordered collections (aka where pairs is non-deterministic or the order is undefined) will result in undefined behaviour. An error counts to undefined behaviour in my opinion so its just a note about documentation.
Btw would we want keys or pairs? I guess keys is already the right way because it sort of separates column specification from actual merging(?)

Reply:
I understand that this is a "yes" to what @nalimilan proposed and what is now on GitHub. I will now review the documentation and add tests to finalize this PR then.